enterprise/backend/src/metabase_enterprise/semantic_search/embedding.clj

```diff
@@ -83,159 +81,140 @@
-          endpoint "https://api.openai.com/v1/embeddings"]
-      (when-not api-key
-        (throw (ex-info "OpenAI API key not configured" {:setting "ee-openai-api-key"})))
-      (try
-        (log/debug "Generating" (count texts) "OpenAI embeddings in batch")
-        (-> (http/post endpoint
-                       {:headers {"Content-Type" "application/json"
-                                  "Authorization" (str "Bearer " api-key)}
-                        :body    (json/encode {:model model
-                                               :input texts})})
-            :body
-            (json/decode true)
-            :data
-            (->> (map :embedding)
-                 vec))
-        (catch Exception e
-          (log/error e "Failed to generate OpenAI embeddings for batch of" (count texts) "texts"
-                     "with token count:" (count-tokens-batch texts))
-          (throw e)))))
-
-  (-pull-model [_]
-    (log/info "OpenAI provider does not require pulling a model"))
-
-  (-model-dimensions [_ model]
-    (get openai-supported-models model)))
-
-(def ^:private providers
-  "Registry of available embedding providers."
-  {:ollama (->OllamaProvider)
-   :openai (->OpenAIProvider)})
-
-(defn get-provider
-  "Get the configured embedding provider according to the ee-embedding-provider setting"
-  []
-  (let [provider-key (keyword (semantic-settings/ee-embedding-provider))]
-    (if-let [provider (get providers provider-key)]
-      provider
-      (throw (ex-info (str "Unknown embedding provider: " provider-key)
-                      {:provider provider-key
-                       :available-providers (keys providers)})))))
-
-(defn get-embedding
-  "Generate a single embedding using the configured provider. Prefer using `process-embeddings-streaming` for bulk index population."
-  [text]
-  (-get-embedding (get-provider) text))
-
-(defn pull-model
-  "Pull/download the model using the configured provider."
+(def ^:dynamic ^:redef *active-model*
+  "The model being used by the current semantic search engine."
+  nil)
+
+(defn get-active-model
+  "Returns *active-model* if defined.
+
+  Otherwise get the environments default embedding model according to the ee-embedding-provider / ee-embedding-model settings.
+
+  Requires the model dimensions are defined in ollama-supported-models or openai-supported-models (throws if not)."
   []
-  (-pull-model (get-provider)))
-
-(defn model-dimensions
-  "Get the number of dimensions for the configured provider and model."
-  ([]
-   (let [provider (get-provider)
-         model (get-model)]
-     (-model-dimensions provider model)))
-  ([model]
-   (-model-dimensions (get-provider) model)))
-
-;; TODO: dedupe embedding fetching for identical values
+  (or *active-model*
+      (let [provider (semantic-settings/ee-embedding-provider)
+            models (case provider
+                     "ollama" ollama-supported-models
+                     "openai" openai-supported-models
```

Comment:
maybe we could just have a `supported-models` multimethod as well to make it even easier to extend to a new provider?